### PR TITLE
librbd/cache/pwl: Fix pmem cache fragment issue

### PR DESCRIPTION
**Pull request by CongMinYin**
commit 1:  Update pmdk version to introduce a new API: pmemobj_defrag(). This API is necessary to solve pmem cache fragmentation issue.

commit 2: Fix pmem cache fragment issue. I/O may hang due to pmem cache fragment issue when blocks are diffrent
in size. Call pmdk API(pmemobj_defrag) to solve.
Fixes: https://tracker.ceph.com/issues/49879

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by CongMinYin** [comment]
Hi @tchaikov , I want to  update pmdk version to v1.10. It is necessary to introduce a new API to  fix bug. The API call here does not involve libpmem2, so there is no problem now. But after the upgrade, if we want to use libpmem2, cmake and dependency need to be change, Yingxin will discuss with you in https://github.com/ceph/ceph/pull/39154.

**Comment by CongMinYin** [comment]
Hi @dillaman , after consulting the pmdk team again, there is a new API(pmdk version>=1.8) to solve this problem. The commit 2 completely solves the fragmentation problem to avoid IO hang.  Pre PR is https://github.com/ceph/ceph/pull/39915. When the cache is empty, there is still no space to allocate. Defragment.

**Comment by CongMinYin** [comment]
jenkins test api

**Review by trociny — APPROVED** [review]
The RBD change LGTM.

I leave it to @tchaikov to approve the pmdk version change.

File: cmake/modules/Buildpmem.cmake [hunk 1]
```diff
@@ -16,7 +16,7 @@ function(build_pmem)
 
   ExternalProject_Add(pmdk_ext
       GIT_REPOSITORY "https://github.com/ceph/pmdk.git"
-      GIT_TAG "1.7"
+      GIT_TAG "1.10"
```
Comment (tchaikov) [inline cmake/modules/Buildpmem.cmake:19]: so we cannot use the libpmem-dev shipped by ubuntu focal anymore..

**Review by tchaikov — APPROVED** [review]
i approve the Buildpmem.cmake change.

**Comment by trociny** [comment]
@CongMinYin shaman failed to build it [1].

On centos it uses the system pmdk, which is of lower version:
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-2755-g9360813c/rpm/el8/BUILD/ceph-17.0.0-2755-g9360813c/src/librbd/cache/pwl/rwl/WriteLog.cc:903:9: error: there are no arguments to 'pmemobj_defrag' that depend on a template parameter, so a declaration of 'pmemobj_defrag' must be available [-fpermissive]
         pmemobj_defrag(m_log_pool, NULL, 0, NULL);
         ^~~~~~~~~~~~~~
```
On ubuntu it uses the pmdk from git but fails when building it:
```
../utils/md2man.sh libpmem/libpmem.7.md default.man libpmem/libpmem.7
../utils/md2man.sh: line 62: pandoc: command not found
```
So pandoc dependency should be added or building the mans disabled somehow.

[1]  https://shaman.ceph.com/builds/ceph/wip-mgolub-testing/9360813c9e441e7fd875c0826b53edb212ec0bab/

File: cmake/modules/Buildpmem.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/Buildpmem.cmake:19]: i created #40578 to enable find_package() to specify the version of libpmem.

**Comment by cyx1231st** [comment]
pandoc can be disabled according to https://github.com/ceph/ceph/pull/39154

File: cmake/modules/Buildpmem.cmake (hunk 1 above)
Comment (CongMinYin) [reply cmake/modules/Buildpmem.cmake:19]: > rsion of libpmem.

As this PR merged, I cherry-pick Yingxin's PR: https://github.com/ceph/ceph/pull/39154. And modified find_package() version, please review again, Thanks.

Comment (CongMinYin) [reply cmake/modules/Buildpmem.cmake:19]: > i created #40578 to enable find_package() to specify the version of libpmem.

As this PR merged, I cherry-pick Yingxin's PR: https://github.com/ceph/ceph/pull/39154. And modified find_package() version, please review again, Thanks.

Comment (tchaikov) [reply cmake/modules/Buildpmem.cmake:19]: i'd suggest update `ceph/ceph.spec.in` to set `_system_pmdk` to 0 accordingly, as, at the time of writing, the only distro which packages libpmem 1.10 is fedora rawhide.

Comment (tchaikov) [reply cmake/modules/Buildpmem.cmake:19]: also, to require pmem libraries v1.10, we will have to either

- include pmdk as a submodule, see #40186
- or include pmdk in `make-dist`.

**Review by tchaikov — CHANGES_REQUESTED** [review]
see my inlined comments.

File: cmake/modules/Buildpmem.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/Buildpmem.cmake:19]: i'd suggest go with the 2nd approach. and use the precompiled packages while running "make check" and packaging for teuthology, see #40617

Comment (CongMinYin) [reply cmake/modules/Buildpmem.cmake:19]: OK, I'll try to do it in `make-dist`. But it may take some time because I'm not familiar with it

Comment (hualongfeng) [reply cmake/modules/Buildpmem.cmake:19]: @CongMinYin @tchaikov  I have finished the 2nd approach, please see #40866

**Comment by CongMinYin** [comment]
@tchaikov Is there anything else to do with pmdk?  If not, could you please check the upgrade on various platforms, because I can only try it on Ubuntu.

**Comment by tchaikov** [comment]
```
[ 84%] Building CXX object src/librbd/CMakeFiles/librbd_plugin_pwl_cache.dir/cache/pwl/rwl/WriteLog.cc.o
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc: In member function 'void librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::reserve_cache(librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::C_BlockIORequestT*, bool&, bool&)':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc:904:9: error: there are no arguments to 'pmemobj_defrag' that depend on a template parameter, so a declaration of 'pmemobj_defrag' must be available [-fpermissive]
         pmemobj_defrag(m_log_pool, NULL, 0, NULL);
         ^~~~~~~~~~~~~~
```

https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-16-1733/f021b9243e67341224411146f0423bb66628ce0a/default/265858/

**Comment by CongMinYin** [comment]
> pmemobj_defrag'



> ```
> [ 84%] Building CXX object src/librbd/CMakeFiles/librbd_plugin_pwl_cache.dir/cache/pwl/rwl/WriteLog.cc.o
> /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc: In member function 'void librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::reserve_cache(librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::C_BlockIORequestT*, bool&, bool&)':
> /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc:904:9: error: there are no arguments to 'pmemobj_defrag' that depend on a template parameter, so a declaration of 'pmemobj_defrag' must be available [-fpermissive]
>          pmemobj_defrag(m_log_pool, NULL, 0, NULL);
>          ^~~~~~~~~~~~~~
> ```
> 
> https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-16-1733/f021b9243e67341224411146f0423bb66628ce0a/default/265858/

I'm sorry to reply so late. Could you tell me if this error occurred in "make check" or "make dist". This looks like ceph didn't linked to the correct version(v1.10) of pmdk in CentOS system, so the function cannot be found. I need to find the difference under CentOS.

**Comment by tchaikov** [comment]
> > pmemobj_defrag'
> 
> > ```
> > [ 84%] Building CXX object src/librbd/CMakeFiles/librbd_plugin_pwl_cache.dir/cache/pwl/rwl/WriteLog.cc.o
> > /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc: In member function 'void librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::reserve_cache(librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::C_BlockIORequestT*, bool&, bool&)':
> > /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc:904:9: error: there are no arguments to 'pmemobj_defrag' that depend on a template parameter, so a declaration of 'pmemobj_defrag' must be available [-fpermissive]
> >          pmemobj_defrag(m_log_pool, NULL, 0, NULL);
> >          ^~~~~~~~~~~~~~
> > ```
> > 
> > 
> > https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-16-1733/f021b9243e67341224411146f0423bb66628ce0a/default/265858/
> 
> I'm sorry to reply so late. Could you tell me if this error occurred in "make check" or "make dist". This looks like ceph didn't linked to the correct version(v1.10) of pmdk in CentOS system, so the function cannot be found. I need to find the difference under CentOS.

please follow the link above. it was neither "make check" nor "make dist". i was trying to build Ceph packages on CentOS from the dist tar ball created using `make dist`.

**Comment by CongMinYin** [comment]
> > > pmemobj_defrag'
> > 
> > 
> > > ```
> > > [ 84%] Building CXX object src/librbd/CMakeFiles/librbd_plugin_pwl_cache.dir/cache/pwl/rwl/WriteLog.cc.o
> > > /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc: In member function 'void librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::reserve_cache(librbd::cache::pwl::rwl::WriteLog<ImageCtxT>::C_BlockIORequestT*, bool&, bool&)':
> > > /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-3302-gf021b924/rpm/el8/BUILD/ceph-17.0.0-3302-gf021b924/src/librbd/cache/pwl/rwl/WriteLog.cc:904:9: error: there are no arguments to 'pmemobj_defrag' that depend on a template parameter, so a declaration of 'pmemobj_defrag' must be available [-fpermissive]
> > >          pmemobj_defrag(m_log_pool, NULL, 0, NULL);
> > >          ^~~~~~~~~~~~~~
> > > ```
> > > 
> > > 
> > > https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-16-1733/f021b9243e67341224411146f0423bb66628ce0a/default/265858/
> > 
> > 
> > I'm sorry to reply so late. Could you tell me if this error occurred in "make check" or "make dist". This looks like ceph didn't linked to the correct version(v1.10) of pmdk in CentOS system, so the function cannot be found. I need to find the difference under CentOS.
> 
> please follow the link above. it was neither "make check" nor "make dist". i was trying to build Ceph packages on CentOS from the dist tar ball created using `make dist`.

Hi @tchaikov , I tried in CentOS.  I think the failure is because the pmdk used in CentOS is 1.6 by default.  If we have to upgrade to 1.10, we will not be able to use the system's pmdk. How do you think?

see  https://github.com/ceph/ceph/pull/40052

**Comment by CongMinYin** [comment]
Hi @tchaikov , I push a new commit to fix pmdk version judgement error. Now, if set WITH_SYSTEM_PMDK in CentOS, do_cmake will report error unsuitable version.

```
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find pmem: Found unsuitable version "1.6.1", but required is at
  least "1.10" (found pmem_pmemobj_INCLUDE_DIR;pmem_pmem_INCLUDE_DIR)
```

File: cmake/modules/Findpmem.cmake [hunk 2]
```diff
@@ -20,7 +20,7 @@ foreach(component ${pmem_FIND_COMPONENTS})
   if(pmem_FIND_VERSION_EXACT)
     set(pkg_module_spec "lib${component}=${pmem_FIND_VERSION}")
   elseif(pmem_FIND_VERSION)
-    set(pkg_module_spec "lib${component}>=${pmem_FIND_VERSION}")
+    set(pkg_module_spec "lib${component}<=${pmem_FIND_VERSION}")
```
Comment (tchaikov) [inline cmake/modules/Findpmem.cmake:23]: i don't understand this change. per https://cmake.org/cmake/help/latest/module/FindPkgConfig.html

> `foo>=3.1` matches any version from 3.1 or later.

Comment (CongMinYin) [reply cmake/modules/Findpmem.cmake:23]: > i don't understand this change. per https://cmake.org/cmake/help/latest/module/FindPkgConfig.html
> 
> > `foo>=3.1` matches any version from 3.1 or later.

pmem_FIND_VERSION is system version(1.6), so it should >= required verison. Corresponding to this example, foo means pmem_FIND_VERSION,  3.1 means our required version.

**Comment by CongMinYin** [comment]
This PR still leaves a part of the work:  set with_system_ pmdk off in `ceph.spec.in`, change it to [WIP].

File: cmake/modules/Findpmem.cmake (hunk 2 above)
Comment (tchaikov) [reply cmake/modules/Findpmem.cmake:23]: no, `pmem_FIND_VERSION` in this case is the requested version string, per https://cmake.org/cmake/help/latest/command/find_package.html#version-selection

> `PACKAGE_FIND_VERSION`
>
>        Full requested version string

i think the correct fix should be #41022

Comment (CongMinYin) [reply cmake/modules/Findpmem.cmake:23]: Thanks for solving the problem and explaining, I will withdraw this commit.

**Comment by tchaikov** [comment]
> This PR still leaves a part of the work: set with_system_ pmdk off in `ceph.spec.in`. changed PR status to [WIP].

i included a change to update `ceph.spec.in` in #41022. @cyx1231st @CongMinYin could you take a look?

**Comment by CongMinYin** [comment]
rebase done.

**Comment by idryomov** [comment]
jenkins test make check

**Review by idryomov — CHANGES_REQUESTED** [review]
The first commit talks about disabling documentation but it looks like it has been done already in one of the previous cmake PRs.  Can you update the commit message?

**Comment by CongMinYin** [comment]
> The first commit talks about disabling documentation but it looks like it has been done already in one of the previous cmake PRs. Can you update the commit message?

done.

**Comment by tchaikov** [comment]
this change builds at https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-27-2204/311c8c37e59b5e7b0a88005ca51d14936edecfa4/

rados test suite (still running): https://pulpito.ceph.com/kchai-2021-04-28_02:38:11-rados-wip-kefu-testing-2021-04-27-2204-distro-basic-smithi/

**Comment by idryomov** [comment]
Does this fully address https://tracker.ceph.com/issues/49879 or is there more work to be done?

**Comment by CongMinYin** [comment]
> Does this fully address https://tracker.ceph.com/issues/49879 or is there more work to be done?

This PR can fully fix this issue.

**Comment by idryomov** [comment]
Good, can you add `Fixes: https://tracker.ceph.com/issues/49879` to the second commit?

**Comment by CongMinYin** [comment]
> Good, can you add `Fixes: https://tracker.ceph.com/issues/49879` to the second commit?

done.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
https://pulpito.ceph.com/dis-2021-04-28_17:10:10-rbd-wip-dis-testing-distro-basic-smithi/

Failure

https://tracker.ceph.com/issues/49876

unrelated.

**Pull request merged**